src/lazy.rs

```diff
@@ -54,3 +58,53 @@ impl LazyBool {
         self.0.unsync_init(|| init() as usize) != 0
     }
 }
+
+// This structure represents a lazily initialized static pointer value.
+///
+/// It's intended to be used for weak linking of a C function that may
+/// or may not be present at runtime.
+///
+/// Based off of the DlsymWeak struct in libstd:
+/// https://github.com/rust-lang/rust/blob/1.61.0/library/std/src/sys/unix/weak.rs#L84
+/// except that the caller must manually cast self.ptr() to a function pointer.
+pub struct LazyPtr {
+    addr: AtomicPtr<c_void>,
+}
+
+impl LazyPtr {
+    /// A non-null pointer value which indicates we are uninitialized.
+    ///
+    /// This constant should ideally not be a valid pointer.
+    /// However, if by chance initialization function passed to the `get`
+    /// method does return UNINIT, there will not be undefined behavior.
+    /// The initialization function will just be called each time `get()`
+    /// is called. This would be inefficient, but correct.
+    const UNINIT: *mut c_void = !0usize as *mut c_void;
```

Comment:
Isn't `!0usize` more likely to be a valid pointer than 1 since at least on some targets the lowest word/byte/page is designed to trap when executed or dereferenced to catch bugs? I think if we want to make a change like this, we should do it in a separate PR.

Reply:
It can not be a valid function pointer since such function will not be able contain even one instruction.

Reply:
SGTM, `!0` makes more sense for "Impossible function pointer value" than `1` (although either ever being used is vanishingly unlikely).